### DISCUSSION TITLE
多维数组会导致IO的方向推断错误，这是不是一个bug?(Is it a bug that multidimensional arrays lead to incorrect inference of the direction of IO?)

### DISCUSSION THREAD

**Discussion by qinjun-li** [chipsalliance/chisel#2038] [Q&A]
这个是我列出来的例子(This is the example I listed):
https://scastie.scala-lang.org/endbwvnsRBG76vRLUTE8rQ

它会报错(This is the example I listed):
```
firrtl.passes.CheckFlows$WrongFlow:  @[main.scala 16:8]: [module Top]  Expression m1.o[0][0].bits is used as a SinkFlow but can only be used as a SourceFlow.
```
实际上方向是没有问题的,但是好像 firrtl 推断错了(In fact, the direction is fine, but it seems that firrtl is inferred incorrectly.).

Translations are directly copy&paste form [DeepL](https://www.deepl.com).

**Comment by sequencer** [comment]
I think this is a bug from `firrtl.passes.CheckFlows`, I'll investigate today.

**Reply by sequencer** [reply]
Here is the minimal reproducible code in FIRRTL, 
https://scastie.scala-lang.org/C8Ck2XSpRKGeDtfW7ygTKg

**Reply by sequencer** [reply]
This is a bug in FIRRTL, bug issue created chipsalliance/firrtl#2293, let's discuss there.

**Reply by seldridge** [reply]
I think this is actually the "Valid has an Output and behaves weirdly" issue? The actual FIRRTL error looks write for the FIRRTL produced. This appears to be Chisel not doing what the user would expect here. If the `Flipped` cast is moved outwards, that will fix it (but there may be more complicated examples where this won't fix it).

Something like: https://scastie.scala-lang.org/I1pKggaRR3iExodBWnAVzQ

**Reply by sequencer** [reply]
Quite strange.
```scala
class Valid[+T <: Data](gen: T) extends Bundle {
  /** A bit that will be asserted when `bits` is valid */
  val valid = Output(Bool())

  /** Some data */
  val bits  = Output(gen)

  /** True when `valid` is asserted
    * @return a Chisel [[Bool]] true if `valid` is asserted
    */
  def fire(dummy: Int = 0): Bool = valid

  override def cloneType: this.type = Valid(gen).asInstanceOf[this.type]
}
```
Seems OK?

**Comment by lsteveol** [comment]
I actually ran into this earlier. I just happend to (_thankfully_) stumble across this issue. Here is another example if needed
https://scastie.scala-lang.org/ELfvxk7VRU2ooDxtRLMofA

Mad props to @seldridge for the suggestion. I can keep what is left of the hair on my head that I haven't pulled out yet!
